src/dist/replication/meta_server/meta_server_failure_detector.cpp

```diff
@@ -263,6 +263,10 @@ void meta_server_failure_detector::on_ping(const fd::beacon_msg& beacon, ::dsn::
     {
         ack.is_master = false;
         ack.primary_node = get_primary();
+        {
```

Comment:
@shengofsun I'm inlining the answers to your questions below.

_- First I want to know the exact meaning of "A non-master meta server not handle the Pings." Is it means that a replica server should never receive a FD message with "ack.is_master == true"? If this is what you mean, I think the logic in "replication_failure_detector::end_ping" can be simplified._

Zhenyu: I mean a non-master meta server should simply ignore any ping messages, or tell the replica who is the right master/leader to connect to. Other than that, it should do nothing, because it is not the mater/leader. The replica can recv messages with both ```ack.is_master == true or false``` in this case.

_- As what I described above, the replica server's state will be influenced if a non-master meta doesn't handle the Ping. This is mainly because how long a meta server switch to a master depends on the distributed_lock_service's implementation. And it may be longer than the lease period. This implies that a replica server will suffer from being inactive as long as a meta server switching happens._

Zhenyu: this is expected and defined by the perfect FD protocol. I understand that you are trying to fix this issue as this seems unnecessary. We may discuss this separately later (or we discussed this before? sorry for my bad memory in that case:)). I think the principle is to not have our invariance in risk which may leads to data inconsistency and/or data lost that is not affordable than being unavailable for a while. The same is applied to your last item below.

_- Another issue is that no matter how M handles Ping from R, the perfect FD rule will be broken during the period that M has become a master but the first beacon from R hasn't arrived. We may need to store the worker list to the remote storage(e.g. zookeeper) to resolve this._

Zhenyu: agree with your solution. There is also an alternative solution that we conservatively think a replica is alive IF (1) the replica is NOT found on the leader meta server; AND (2) the leader meta server just becomes the leader within the grace (or lease?) period. This avoids saving the machine state in meta store. 

_- Well, the next tricky thing in our project is that the breaking of perfect FD is not so serious. The replica group won't be inconsistent even if two primaries exist, as a preparing message from an old primary will be rejected by the new primary. So I guess we can just keep this bug here for the time being?_

Zhenyu: this is true only with certain configurations (e.g., min-replica-count-for-write >= majority of the replica degree). However, we don't really test this seriously, and there are always settings that can lead to inconsistent data committed at the same decree (e.g., when min-replica-count-for-write == 2 and replica degree == 4).